src/librustc/ty/mod.rs

```diff
@@ -566,38 +568,110 @@ impl <'gcx: 'tcx, 'tcx> Canonicalize<'gcx, 'tcx> for Ty<'tcx> {
     }
 }
 
+extern {
+    /// A dummy type used to force Slice to by unsized without requiring fat pointers
+    type OpaqueSliceContents;
+}
+
 /// A wrapper for slices with the additional invariant
 /// that the slice is interned and no other slice with
 /// the same contents can exist in the same context.
 /// This means we can use pointer + length for both
 /// equality comparisons and hashing.
-#[derive(Debug, RustcEncodable)]
-pub struct Slice<T>([T]);
+/// Empty slices are encoded in the pointer to this as `1`
+pub struct Slice<T>(PhantomData<T>, OpaqueSliceContents);
+
+const EMPTY_SLICE: usize = 1;
+
+impl<T> Slice<T> {
+    /// Returns the offset of the array
+    #[inline(always)]
+    fn offset() -> usize {
+        // Align up the size of the len (usize) field
+        let align = mem::align_of::<T>();
+        let align_mask = align - 1;
+        let offset = mem::size_of::<usize>();
+        (offset + align_mask) & !align_mask
+    }
+}
+
+impl<T: Copy> Slice<T> {
+    #[inline]
+    fn from_arena<'tcx>(arena: &'tcx SyncDroplessArena, slice: &[T]) -> &'tcx Slice<T> {
+        assert!(!mem::needs_drop::<T>());
+        assert!(mem::size_of::<T>() != 0);
+        assert!(slice.len() != 0);
+
+        let offset = Slice::<T>::offset();
+        let size = offset + slice.len() * mem::size_of::<T>();
+
+        let mem: *mut u8 = arena.alloc_raw(
+            size,
+            cmp::max(mem::align_of::<T>(), mem::align_of::<usize>())).as_mut_ptr();
+
+        unsafe {
+            // Write the length
+            *(mem as *mut usize) = slice.len();
+
+            // Write the elements
+            let arena_slice = slice::from_raw_parts_mut(
+                mem.offset(offset as isize) as *mut T,
+                slice.len());
+            arena_slice.copy_from_slice(slice);
+
+            &*(mem as *const Slice<T>)
+        }
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for Slice<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        (**self).fmt(f)
+    }
+}
+
+impl<T: Encodable> Encodable for Slice<T> {
+    #[inline]
+    fn encode<S: Encoder>(&self, s: &mut S) -> Result<(), S::Error> {
+        (**self).encode(s)
+    }
+}
 
 impl<T> PartialEq for Slice<T> {
     #[inline]
     fn eq(&self, other: &Slice<T>) -> bool {
-        (&self.0 as *const [T]) == (&other.0 as *const [T])
+        (self as *const _) == (other as *const _)
```

Comment:
`PartialEq` and `Hash` don't agree. Is it possible that we ever have two slices with the same address but different lengths?

Reply:
That's not possible as the length is stored at that address.

Reply:
Yeah.

Reply:
The `Hash` implementation should just hash the pointer then. Should be faster too.

Reply:
The `HashStable` impl in `impls_ty.rs` should also just use the pointer as the cache key.